monix-catnap/shared/src/main/scala/monix/catnap/CancelableF.scala

```diff
@@ -0,0 +1,158 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package monix.catnap
+
+import cats.Applicative
+import cats.effect.{CancelToken, Sync}
+import monix.catnap.cancelables.BooleanCancelableF
+import monix.execution.annotations.UnsafeBecauseImpure
+import monix.execution.internal.Platform
+
+import scala.collection.mutable.ListBuffer
+
+/** Represents a pure data structure that describes an effectful,
+  * idempotent action that can be used to cancel async computations,
+  * or to release resources.
+  *
+  * This is the pure, higher-kinded equivalent of
+  * [[monix.execution.Cancelable]] and can be used in combination
+  * with data types meant for managing effects, like
+  * `Task`, `Coeval` or `cats.effect.IO`.
+  *
+  * Note: the `F` suffix comes from this data type being abstracted
+  * over `F[_]`.
+  */
+trait CancelableF[F[_]] {
+  def cancel: CancelToken[F]
+}
+
+object CancelableF {
+  /**
+    * Given a token that does not guarantee idempotency, wraps it
+    * in a [[CancelableF]] value that guarantees the given token
+    * will execute only once.
+    */
+  def apply[F[_]](token: F[Unit])(implicit F: Sync[F]): F[CancelableF[F]] =
+    BooleanCancelableF(token).asInstanceOf[F[CancelableF[F]]]
+
+  /**
+    * Unsafe version of [[apply]].
+    *
+    * This function is unsafe because creating the returned
+    * [[BooleanCancelableF]] allocates internal shared mutable
+    * state, thus breaking referential transparency, which can
+    * catch users by surprise.
+    */
+  @UnsafeBecauseImpure
+  def unsafeApply[F[_]](token: F[Unit])(implicit F: Sync[F]): CancelableF[F] =
+    BooleanCancelableF.unsafeApply(token)
+
+  /**
+    * Wraps a cancellation token into a [[CancelableF]] instance.
+    *
+    * Compared with [[apply]] the returned value does not
+    * guarantee idempotency.
+    *
+    * N.B. the returned result is as pure as the wrapped result.
+    * Since we aren't allocating mutable internal state for creating
+    * this value, we don't need to return the value in `F[_]`,
+    * like in [[apply]].
+    */
+  def wrap[F[_]](token: CancelToken[F]): CancelableF[F] =
+    new CancelableF[F] { def cancel = token }
+
+  /**
+    * Returns a dummy [[CancelableF]] that doesn't do anything.
+    */
+  def empty[F[_]](implicit F: Applicative[F]): CancelableF[F] =
+    new CancelableF[F] { def cancel = F.unit }
+
+  /** Builds a [[CancelableF]] reference from a sequence,
+    * cancelling everything when `cancel` gets evaluated.
+    */
+  def collection[F[_]](refs: CancelableF[F]*)(implicit F: Sync[F]): CancelableF[F] =
+    unsafeApply[F](cancelAll(refs:_*))
```

Comment:
Should this not be `@UnsafeBecauseImpure`?

Reply:
The actual implementation has no state. Should it be?

Reply:
I mean, if it's unsafe, I'd rather wrap it in `F[_]`, but figured there's no point in doing it.

Not sure if I'm right. Maybe I should just wrap it in `F[_]` to be on the safe side.

Reply:
I thought the `unsafeApply` allocates mutable state, no?

Reply:
Ugh, yes, that's a good point. I think my intention was to use `wrap`, because `CancelableF` references have to ensure idempotency on their own, no extra protections needed for such a "collection".

Thanks for pointing it out, definitely needs a fix.

Reply:
Glad it helped 👍